src/workerd/api/streams/readable-source.h

```diff
@@ -0,0 +1,214 @@
+#pragma once
+
+#include <workerd/io/worker-interface.capnp.h>
+
+#include <kj/debug.h>
+
+namespace kj {
+class AsyncInputStream;
+}
+
+namespace workerd {
+
+class IoContext;
+namespace api {
+template <typename T>
+struct DeferredProxy;
+}  // namespace api
+
+namespace jsg {
+class Lock;
+}  // namespace jsg
+
+namespace api::streams {
+
+class WritableStreamSink;
+
+// A ReadableStreamSource is primarily intended to serve as a bridge between kj::AsyncInputStream
+// and the ReadableStream API. However, it can also be used directly by KJ-space code that needs
+// deferred proxying. While ReadableStreamSource should probably have been a more JS-friendly
+// API, it's a bit too late to change that now. Use the ReadableStreamSourceJsAdapter in the
+// readable-source-adapter.h file to wrap a ReadableStreamSource for use from JavaScript.
+//
+// Not all ReadableStreamSource implementations will be explicitly backed by a KJ stream;
+// some might be test implementations that generate data on the fly, for instance.
+//
+// A ReadableStreamSource must be treated like a KJ I/O object. Instances that are held
+// by any JS-heap objects must be held by an IoOwn.
+//
+// If the ReadableStreamSource is canceled or dropped, all pending read() and pumpTo()
+// promises will be canceled.
+//
+// Only one read() or pumpTo() operation may be pending at a time. Attempting to
+// initiate a second read() or pumpTo() while one is already pending will result in
+// a rejected promise.
+//
+// It is **NOT** intended that you should implement this interface for general use.
+// It is only intended to be implemented by specific classes within workerd for the
+// purpose of bridging between kj/js streams. Streams that operate at the kj level
+// should implement the kj::Async*Stream interfaces, and streams that operate at
+// the JS level should implement to the UnderlyingSource interface. This is a
+// departure from what we've done previously but as part of the effort to simplify
+// the streams code, the goal is to reduce the number of different stream interfaces
+// that we implement to.
+class ReadableStreamSource {
```

Comment:
I thought ReadableStreamSource currently lives in streams/common.h, but don't see it being deleted in this PR. Does that mean there are two classes of the same name somehow? Or am I confused...

Reply:
Implementing the replacement + tests first, then will work on swapping it in. Intentionally going slow and incremental on this given that it's so functionally critical to remain stable.